File: app/src/main/java/com/example/dreamloaf/adapter/ProductAdapter.java
package com.example.dreamloaf.adapter;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.dreamloaf.R;
import com.example.dreamloaf.data.Product;

public class ProductAdapter extends ListAdapter<Product, ProductAdapter.ProductViewHolder> {

    public interface OnDeleteClickListener {
        void onDeleteClick(Product product);
    }

    public interface OnItemClickListener {
        void onItemClick(Product product);
    }

    private OnDeleteClickListener onDeleteClickListener;
    private OnItemClickListener onItemClickListener;

    private static final DiffUtil.ItemCallback<Product> DIFF_CALLBACK = new DiffUtil.ItemCallback<Product>() {
        @Override
        public boolean areItemsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @SuppressLint("DiffUtilEquals")
        @Override
        public boolean areContentsTheSame(@NonNull Product oldItem, @NonNull Product newItem) {
            return oldItem.equals(newItem);
        }
    };

    public ProductAdapter() {
        super(DIFF_CALLBACK);
    }

    public void setOnDeleteClickListener(OnDeleteClickListener listener) {
        this.onDeleteClickListener = listener;
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.onItemClickListener = listener;
    }

    @NonNull
    @Override
    public ProductViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_product, parent, false);
        return new ProductViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ProductViewHolder holder, int position) {
        Product product = getItem(position);
        holder.bind(product);
    }

    class ProductViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvName;
        private final TextView tvWeight;
        private final TextView tvPrice;
        private final View btnDelete;

        public ProductViewHolder(@NonNull View itemView) {
            super(itemView);
            tvName = itemView.findViewById(R.id.tvName);
            tvWeight = itemView.findViewById(R.id.tvWeight);
            tvPrice = itemView.findViewById(R.id.tvPrice);
            btnDelete = itemView.findViewById(R.id.btnDelete);

            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (position != RecyclerView.NO_POSITION && onItemClickListener != null) {
                    Product product = getItem(position);

                    if (v.getTag() != null && (long)v.getTag() + 500 > System.currentTimeMillis()) {
                        onItemClickListener.onItemClick(product);
                    }
                    v.setTag(System.currentTimeMillis());
                }
            });
        }

        public void bind(Product product) {
            tvName.setText(product.getName());
            tvWeight.setText(String.format("%.1f г", product.getWeight()));
            tvPrice.setText(String.format("%.2f руб.", product.getPrice()));

            btnDelete.setOnClickListener(v -> {
                if (onDeleteClickListener != null) {
                    onDeleteClickListener.onDeleteClick(product);
                }
            });
        }
    }
}

File: app/src/main/java/com/example/dreamloaf/adapter/ProductAdapter.kt
package com.example.dreamloaf.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.dreamloaf.R
import com.example.dreamloaf.adapter.ProductAdapter.ProductViewHolder
import com.example.dreamloaf.data.Product

class ProductAdapter : androidx.recyclerview.widget.ListAdapter<Product, ProductViewHolder?>(
    DIFF_CALLBACK
) {
    interface OnDeleteClickListener {
        fun onDeleteClick(product: Product?)
    }

    interface OnItemClickListener {
        fun onItemClick(product: Product?)
    }

    private var onDeleteClickListener: OnDeleteClickListener? = null
    private var onItemClickListener: OnItemClickListener? = null

    fun setOnDeleteClickListener(listener: OnDeleteClickListener?) {
        this.onDeleteClickListener = listener
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.onItemClickListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_product, parent, false)
        return ProductViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ProductViewHolder, position: Int) {
        val product: Product = getItem(position)
        holder.bind(product)
    }

    public inner class ProductViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvName: TextView
        private val tvWeight: TextView
        private val tvPrice: TextView
        private val btnDelete: View

        init {
            tvName = itemView.findViewById<TextView>(R.id.tvName)
            tvWeight = itemView.findViewById<TextView>(R.id.tvWeight)
            tvPrice = itemView.findViewById<TextView>(R.id.tvPrice)
            btnDelete = itemView.findViewById<View>(R.id.btnDelete)

            itemView.setOnClickListener(View.OnClickListener { v: View? ->
                val position: Int = getAdapterPosition()
                if (position != RecyclerView.NO_POSITION && onItemClickListener != null) {
                    val product: Product? = getItem(position)

                    if (v!!.getTag() != null && v.getTag() as Long + 500 > System.currentTimeMillis()) {
                        onItemClickListener!!.onItemClick(product)
                    }
                    v.setTag(System.currentTimeMillis())
                }
            })
        }

        fun bind(product: Product) {
            tvName.text = product.name
            tvWeight.text = String.format("%.1f г", product.weight)
            tvPrice.text = String.format("%.2f руб.", product.price)
            btnDelete.setOnClickListener { v: View? ->
                onDeleteClickListener?.onDeleteClick(product)
            }
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Product> =
            object : DiffUtil.ItemCallback<Product>() {
                override fun areItemsTheSame(oldItem: Product, newItem: Product): Boolean {
                    return oldItem.id == newItem.id
                }

                @SuppressLint("DiffUtilEquals")
                override fun areContentsTheSame(oldItem: Product, newItem: Product): Boolean {
                    return oldItem == newItem
                }
            }
    }
}